### ISSUE TITLE
Monoid1 class

### ISSUE THREAD

**Issue by markandrus** [fantasyland/fantasy-land#102]
How should a container which forms a monoid when it contains an instance of a monoid be represented? In the included example, we have

```js
// Monoid (value must also be a Monoid)
Id.prototype.empty = function() {
    return new Id(this.value.empty ? this.value.empty() : this.value.constructor.empty());
};
```

But ideally we could put `empty` on `Id` and not the prototype. I wonder if it makes sense to have a Monoid1 class which defines `empty1`, e.g.

```js
Id.empty1 = function(empty) {
  return new Id(empty());
};
```

**Comment by bergus** [comment]
I think you'd need parametric types for that, and you'll have to explicitly instantiate those in JS. Like
```
function makeId(Contained) {
    function Id() { … }
    Id.empty = function() {
        new Id(Contained.empty());
    };
    …
    return Id;
}
var MyId = makeId(myMonoid);
```

**Comment by SimonRichardson** [comment]
Isn't this a very similar argument to #88

**Comment by CrossEye** [comment]
In #88, I started arguing that allowing the definition on both the instance and the constructor (for both `of` and `empty`) was a problem, initially believing that just defining on the instance was the right thing.  But eventually I came to the realization that better would be to define only on the constructor.

So this overlaps that issue regarding

> But ideally we could put empty on Id and not the prototype

But I think @bergus is on the right track for the fundamental concern raised here.
